[dynamo] show stack above dynamo in graph break user tracebacks

williamwen42 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #148736
* __->__ #148401

Also show the line of code relevant to a dynamo-compiled frame, instead of just the first line (this was broken for data-dependent jump graph breaks and for 3.11+).

Also collapses resume frames together (use config.verbose to see full stack trace - for developers).

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames

williamwen42 committed:
[dynamo] show stack above dynamo in graph break user tracebacks

[ghstack-poisoned]

williamwen42 committed:
Update on "[dynamo] show stack above dynamo in graph break user tracebacks"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

williamwen42 committed:
Update on "[dynamo] show stack above dynamo in graph break user tracebacks"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

williamwen42 committed:
Update on "[dynamo] show stack above dynamo in graph break user tracebacks"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

zou3519 commented on test/dynamo/test_error_messages.py:
```diff
@@ -930,6 +940,93 @@ def gn():
+  File "test_error_messages.py", line N, in hn
+    torch._dynamo.graph_break()
+
+========== `torch.compile` tracing started here ==========
```
I'm confused about what I'm looking at here

zou3519 replied:
Yeah I don't get what "`torch.compile` tracing started here" is and what is below that line

williamwen42 replied:
Technical answer: everything above this line is what happens before dynamo eval_frame/convert_frame. Everything below this line is what dynamo actually traces.

Further explanation (likely need to work on improving this explanation to users): This part of the stack trace
```
File "test_error_messages.py", line N, in fn
    gn(x + 1)
  File "test_error_messages.py", line N, in gn
    hn(x + 1)
  File "test_error_messages.py", line N, in hn
    torch._dynamo.graph_break()
```
was traced by a previous invocation of dynamo's convert_frame. In this current invocation, `fn, gn, hn` are actually the optimized code objects produced by dynamo (perhaps we should change their function names to reflect this). This convert_frame invocation actually starts tracing from `torch_dynamo_resume_in_hn_at_N`, so the logs reflect this. This behavior happens as a result of dynamo's graph break behavior.

williamwen42 committed:
Update on "[dynamo] show stack above dynamo in graph break user tracebacks"


Also show the line of code relevant to a dynamo-compiled frame, instead of just the first line (this was broken for data-dependent jump graph breaks and for 3.11+).

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

zou3519 commented on test/dynamo/test_exc.py:
```diff
@@ -183,6 +183,11 @@ def fn001(x):
   Developer debug context: Called `torch._dynamo.graph_break()` with args `[]`, kwargs `{}`
 
 User code traceback:
+  File "test_exc.py", line N, in test_graph_break_log
+    torch.compile(fn001, backend="eager")(torch.randn(1))
+
+========== `torch.compile` tracing started here ==========
+
```
Can this be "collapsed" somehow? I would have expected something like:
```
  File "test_exc.py", line N, in test_graph_break_log
    torch.compile(fn001, backend="eager")(torch.randn(1))
  File "test_exc.py", line N, in fn001
    return fn002(x)
  File "test_exc.py", line N, in fn002
    torch._dynamo.graph_break()
```

williamwen42 replied:
Forgot to update this test aha

zou3519 commented on torch/_dynamo/symbolic_convert.py:
```diff
@@ -499,16 +500,31 @@ def log_graph_break(code_options, reason="", exc_info=False, user_stack=None):
+    if config.verbose:
+        user_stack_trace += (
+            f"{stack_above_dynamo_formatted}\n"
+            "========== `torch.compile` tracing started here ==========\n\n"
```
If you're keeping the split for verbose, then we should explain this better. It's okay to say things about resume frames. The user might not understand what is going on so this explanation is more for developers.

zou3519 reviewed (commented):
this is cool! I really like the idea to collapse resume frames

williamwen42 committed:
Update on "[dynamo] show stack above dynamo in graph break user tracebacks"


Also show the line of code relevant to a dynamo-compiled frame, instead of just the first line (this was broken for data-dependent jump graph breaks and for 3.11+).

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

williamwen42 commented:
@pytorchbot merge -i

williamwen42 commented:
@pytorchbot -i

williamwen42 commented:
@pytorchbot merge -i
